### PR TITLE
Add FIPS mode secret

### PR DESCRIPTION
**Pull request by umohnani8**
If the host is in FIPS mode and /etc/system-fips exists
/run/secrets/system-fips is created in the container so that
the container can run in FIPS mode as well.

Moving the changes of https://github.com/projectatomic/buildah/pull/603 here. The secrets pkg will be vendored into buildah and cri-o to reduce code duplication.

Signed-off-by: umohnani8 <umohnani@redhat.com>

### CODE REVIEW

**Comment by umohnani8** [comment]
@rhatdan @mheon PTAL

File: pkg/secrets/secrets.go [hunk 1]
```diff
@@ -187,9 +183,63 @@ func SecretMounts(filePath, mountLabel, containerWorkingDir string) ([]rspec.Mou
...
+	}
 	return mounts, nil
 }
 
+// addFIPSModeSecret creates /run/secrets/system-fips in the container
+// root filesystem if /etc/system-fips exists on hosts.
+// This enables the container to be FIPS compliant and run openssl in
+// FIPS mode as the host is also in FIPS mode.
+func addFIPSsModeSecret(mounts *[]rspec.Mount, containerWorkingDir string) error {
+	secretsDir := "/run/secrets"
```
Comment (mheon) [inline pkg/secrets/secrets.go]: We're going to mount this RW into all containers, which makes me think we should probably not only have one copy?

Comment (mheon) [reply pkg/secrets/secrets.go]: Also, I don't like having this dir non-configurable. For libpod, I'd like to tuck it inside our tmp directory somewhere.

**Comment by mheon** [comment]
Do we need to have a separate set of RO bind mounts for this? That would not be hard to do, but the secrets code would have to sort them into RO/RW for us to do that.

File: pkg/secrets/secrets.go (hunk 1 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: Tuck the "/run/secrets" dir inside a tmp directory or the "system-fips" file. The system-fips file needs to be in /run/secrets in the container as that is what a symlink in the image will be pointing to. If the symlink is not dangling, that means the host was in fips mode so the container should run in fips mode as well.

**Comment by umohnani8** [comment]
Adds functionality for https://trello.com/c/UawSMhdW/572-openssl-fips-mode-in-containers on the libpod side.

File: pkg/secrets/secrets.go (hunk 1 above)
Comment (mheon) [reply pkg/secrets/secrets.go]: `/run/secrets` should be configurable, I would like to put that in a `secrets` dir under libpod's tmp dir

Comment (umohnani8) [reply pkg/secrets/secrets.go]: Makes sense, but this same patch is going to be vendored in to buildah and cri-o. Plus @rhatdan said to have it in `/run/secrets`. Should we make /run/secrets such that the user can't overwrite it?

Comment (rhatdan) [reply pkg/secrets/secrets.go]: /run/secrets is standard, inside the container.  RHEL containers are hard coded to look in /run/secrets in the container.  It can not be configurable.

File: pkg/secrets/secrets.go [hunk 2]
```diff
@@ -187,9 +183,63 @@ func SecretMounts(filePath, mountLabel, containerWorkingDir string) ([]rspec.Mou
 
 		mounts = append(mounts, m)
 	}
+
+	// Add FIPS mode secret
+	// The filePath check is to ensure that this secret is created only once
+	// as SecretMounts is called twice for the two mounts.conf files ("OverrideMountsFile" and "DefaultMountsFile") in the respective order
+	// therefore we want to add the FIPS mode secret, if it exists, after the secrets from the DefaultMountsFile are mounted
+	_, err := os.Stat("/etc/system-fips")
```
Comment (rhatdan) [inline pkg/secrets/secrets.go]: Should check if err != NotExists and return an error.

File: pkg/secrets/secrets.go [hunk 3]
```diff
@@ -187,9 +183,63 @@ func SecretMounts(filePath, mountLabel, containerWorkingDir string) ([]rspec.Mou
...
+	if err == nil && filePath == DefaultMountsFile {
```
Comment (rhatdan) [inline pkg/secrets/secrets.go]: I think fips-mode should be set whether or not the User overrode the default mounts file.

**Comment by rhatdan** [comment]
I guess I could see making these directories read/only if the container is running read-only.  But traditionally we have allowed the secrets directory to be modified by the container, similarly to how we allow the /etc/resolv.conf and /etc/hosts to be modified.

File: pkg/secrets/secrets.go (hunk 2 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: Why return an error? We are just checking if the host has the "system-fips" file, if it doesn't we move on. I don't see why we should fail.

File: pkg/secrets/secrets.go (hunk 3 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: Right, it is always being set. This if code just makes sure this block of code is not called twice for no reason. The secrets code checks both the Override and Default Mounts files.

File: pkg/secrets/secrets.go [hunk 4]
```diff
@@ -187,9 +183,63 @@ func SecretMounts(filePath, mountLabel, containerWorkingDir string) ([]rspec.Mou
...
+	ctrDirOnHost := filepath.Join(containerWorkingDir, secretsDir)
+	if _, err := os.Stat(ctrDirOnHost); os.IsNotExist(err) {
+		if err = os.MkdirAll(ctrDirOnHost, 0755); err != nil {
+			return errors.Wrapf(err, "making container directory on host failed")
+		}
+	}
+	fipsFile := filepath.Join(ctrDirOnHost, "system-fips")
+	// In the event of restart, it is possible for the FIPS mode file to already exist
+	if _, err := os.Stat(fipsFile); os.IsNotExist(err) {
+		if _, err := os.Create(fipsFile); err != nil {
```
Comment (mheon) [inline pkg/secrets/secrets.go]: Need to close the file you create here

File: pkg/secrets/secrets.go (hunk 1 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: @rhatdan if the user volume mounts something to "/run/secrets" using the --volume flag, it overwrites any secrets that were mounted. Should this behaviour stay or should we stop the user from being ale to overwrite "/run/secrets"?
When docker run sees a duplicate mount destination, it errors out telling the user "destination" is duplicated.

File: pkg/secrets/secrets.go [hunk 5]
```diff
@@ -187,9 +183,63 @@ func SecretMounts(filePath, mountLabel, containerWorkingDir string) ([]rspec.Mou
 
 		mounts = append(mounts, m)
 	}
+
+	// Add FIPS mode secret
```
Comment (mheon) [inline pkg/secrets/secrets.go]: Instead of calling this from `SecretMounts()` and needing to check to make sure we aren't adding two FIPS secrets, can you make `AddFIPSModeSecret()` public and call it directly from libpod after we get the normal secrets?

File: pkg/secrets/secrets.go (hunk 1 above)
Comment (mheon) [reply pkg/secrets/secrets.go]: @umohnani8 I like our behavior of not erroring, but we should probably log that a mount is being ignored if we don't add it because of overlap. I don't think we do that now.

File: pkg/secrets/secrets.go (hunk 5 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: That is what I was initially doing, but I spoke with @rhatdan and he pointed out that all secrets mounting should happen in one function, so it is just one thing we need to call to get all the secrets into the container. Thus the code is under `SecretMounts`, which makes sense to me.
Actually I can rework the code a bit to clean it up.

File: pkg/secrets/secrets.go (hunk 1 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: @mheon okay, agreed.

File: pkg/secrets/secrets.go (hunk 4 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: fixed

File: pkg/secrets/secrets.go (hunk 5 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: Reworked it. PTAL

**Comment by umohnani8** [comment]
> I guess I could see making these directories read/only if the container is running read-only. But traditionally we have allowed the secrets directory to be modified by the container, similarly to how we allow the /etc/resolv.conf and /etc/hosts to be modified.

@rhatdan if this is the case, then shouldn't `/run/secrets` be a tmpfs mount as well for the read-only case to allow the container to be able to modify it? If it is not, the containers will fail when it tries to volume mount the secrets.

**Comment by rhatdan** [comment]
@umohnani8 Could you try this out and see what happens?

**Comment by umohnani8** [comment]
@rhatdan yup will do. This looks ready to merge, PTAL.

File: pkg/secrets/secrets.go [hunk 6]
```diff
@@ -127,9 +126,33 @@ func getMountsMap(path string) (string, string, error) {
...
 
-// SecretMounts copies the contents of host directory to container directory
+// SecretMounts copies, adds, and mounts the secrets to the container root filesystem
+func SecretMounts(mountLabel, containerWorkingDir string) ([]rspec.Mount, error) {
+	var secretMounts []rspec.Mount
+	// Add secrets from paths given in the mounts.conf files
+	for _, file := range []string{OverrideMountsFile, DefaultMountsFile} {
+		mounts, err := addSecretsFromMountsFile(file, mountLabel, containerWorkingDir)
+		if err != nil {
+			logrus.Warn("error mounting secrets, skipping...")
```
Comment (mheon) [inline pkg/secrets/secrets.go]: Should print the error in the logrus.Warn so we can debug when things are going bad

File: pkg/secrets/secrets.go [hunk 7]
```diff
@@ -127,9 +126,33 @@ func getMountsMap(path string) (string, string, error) {
...
+	}
+
+	// Add FIPS mode secret if /etc/system-fips exists on the host
+	_, err := os.Stat("/etc/system-fips")
+	if err == nil {
+		if err := addFIPSsModeSecret(&secretMounts, containerWorkingDir); err != nil {
+			return nil, errors.Wrapf(err, "error adding FIPS mode secret to container")
+		}
+	} else if os.IsNotExist(err) {
+		logrus.Warn("/etc/system-fips does not exist on host, not mounting FIPS mode secret")
```
Comment (mheon) [inline pkg/secrets/secrets.go]: Drop this to a Debug. I don't think this file exists unless you have FIPS mode enabled, and the vast majority of systems do not have FIPS mode enabled. We shouldn't warn on something this common.

File: pkg/secrets/secrets.go [hunk 8]
```diff
@@ -127,9 +126,33 @@ func getMountsMap(path string) (string, string, error) {
 	return "", "", errors.Errorf("unable to get host and container dir")
 }
 
-// SecretMounts copies the contents of host directory to container directory
+// SecretMounts copies, adds, and mounts the secrets to the container root filesystem
+func SecretMounts(mountLabel, containerWorkingDir string) ([]rspec.Mount, error) {
```
Comment (mheon) [inline pkg/secrets/secrets.go]: This function never returns an error. I think it's probably worth returning the errors we get back if we fail to mount something, though failures should not be fatal.

Comment (mheon) [reply pkg/secrets/secrets.go]: Correction - it looks like adding the FIPS mode secret causes an error, but the default and override mounts files don't. Can we store errors from those two, so we don't fail immediately if one of them fails but we still return an error at the end?

File: pkg/secrets/secrets.go (hunk 6 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: fixed

File: pkg/secrets/secrets.go (hunk 7 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: fixed

File: libpod/container_internal.go [hunk 9]
```diff
@@ -932,6 +925,8 @@ func (c *Container) generateSpec(ctx context.Context) (*spec.Spec, error) {
 		}
 		if !MountExists(g.Mounts(), dstPath) {
 			g.AddMount(newMount)
+		} else {
+			logrus.Warnf("mount being overwritten by user %q", dstPath)
```
Comment (mheon) [inline libpod/container_internal.go]: Can you rephrase this to make it a little more clear? Maybe "User mount overriding libpod mount at %s"?

File: pkg/secrets/secrets.go (hunk 8 above)
Comment (umohnani8) [reply pkg/secrets/secrets.go]: It returns an error when it fails to add the FIPS mode secret if /etc/system-fips exists on the host. The other errors end up being logged.

Comment (umohnani8) [reply pkg/secrets/secrets.go]: fixed

File: libpod/container_internal.go (hunk 9 above)
Comment (umohnani8) [reply libpod/container_internal.go]: fixed

**Comment by mheon** [comment]
LGTM.
@rhatdan @baude @TomSweeneyRedHat PTAL

**Comment by mheon** [comment]
bot, retest this please

File: pkg/secrets/secrets.go [hunk 10]
```diff
@@ -190,6 +210,51 @@ func SecretMounts(filePath, mountLabel, containerWorkingDir string) ([]rspec.Mou
...
 
+// addFIPSModeSecret creates /run/secrets/system-fips in the container
+// root filesystem if /etc/system-fips exists on hosts.
+// This enables the container to be FIPS compliant and run openssl in
+// FIPS mode as the host is also in FIPS mode.
+func addFIPSsModeSecret(mounts *[]rspec.Mount, containerWorkingDir string) error {
+	secretsDir := "/run/secrets"
+	ctrDirOnHost := filepath.Join(containerWorkingDir, secretsDir)
+	if _, err := os.Stat(ctrDirOnHost); os.IsNotExist(err) {
+		if err = os.MkdirAll(ctrDirOnHost, 0755); err != nil {
```
Comment (TomSweeneyRedHat) [inline pkg/secrets/secrets.go]: Just wondering if there's any value to adding a variable to libpod.conf to allow the user the toggle the perm level.

Comment (rhatdan) [reply pkg/secrets/secrets.go]: Not for this PR.

File: pkg/secrets/secrets.go [hunk 11]
```diff
@@ -127,9 +126,33 @@ func getMountsMap(path string) (string, string, error) {
...
+	for _, file := range []string{OverrideMountsFile, DefaultMountsFile} {
+		mounts, err := addSecretsFromMountsFile(file, mountLabel, containerWorkingDir)
+		if err != nil {
+			logrus.Warnf("error mounting secrets, skipping: %v", err)
+		}
+		secretMounts = append(secretMounts, mounts...)
+	}
+
+	// Add FIPS mode secret if /etc/system-fips exists on the host
+	_, err := os.Stat("/etc/system-fips")
```
Comment (rhatdan) [inline pkg/secrets/secrets.go]: If I can't check if the /etc/system-fips file exists. then I can not tell whether to put the system into fips mode, but we don't even exist.  We should at least logrus.Errorf() this fact,

Comment (mheon) [reply pkg/secrets/secrets.go]: If it's os.IsNotExist(), the system is not in FIPS mode, we should just debugf. If it's an actual access error, I think that's worthy of an Errorf (or returning an errors.Wrapf?)

Comment (umohnani8) [reply pkg/secrets/secrets.go]: Fixed.

Comment (rhatdan) [reply pkg/secrets/secrets.go]: If you can't search /etc then you don't know if the system is in fips mode, so we really should not continue.  (Highly unlikely).

**Comment by rhatdan** [comment]
LGTM
@rh-atomic-bot r+

**Comment by rhatdan** [comment]
@umohnani8 Now we need the same code in BUILDAH and CRI-O.

**Comment by rhatdan** [comment]
HOMU seems to be hung.  Merging manually.
